[Feature]#265 소셜로그인 카카오

gyogyo0212 opened the pull request:
## 📌 개요
소셜로그인 카카오

## ✅ 작업 사항
- [x] 소셜 로그인 서비스 인터페이스(SocialOAuthLoginService) 및 구현체(KakaoOAuthLoginService) 생성

- [x] 카카오 OAuth 사용자 정보(access token 이용) 조회 및 User 엔티티 생성/조회 로직 구현

- [x] JWT 토큰 생성 로직 연동 (access token, refresh token 발급)

- [x] Redis를 이용한 Refresh Token 저장 서비스(RedisTokenService) 구현

- [x] 소셜 로그인 컨트롤러(SocialLoginController)에서 카카오 로그인 서비스 호출 구현

- [x] 예외 처리 및 커스텀 예외(CustomException, ErrorCode) 적용 완료

- [x] JwtProvider 내 토큰 생성 메서드(createAccessToken, createRefreshToken) 수정 완료

- [x] 카카오 개발자 센터 설정

- [x] Redirect URI 등록 (http://localhost:8080/api/auth/kakao/callback)

- [x] REST API 키, 시크릿 키 확인 및 환경변수 등록

- [x] 인가 코드 방식으로 Access Token 발급 기능 구현

- [x] 카카오로부터 인가 코드 받아 백엔드에서 access token 요청

- [x] KakaoOAuthLoginService에 인가 코드 -> Access Token 요청 메서드 추가

- [x] 카카오 로그인 콜백 API 엔드포인트 구현

- [x] /api/auth/kakao/callback 컨트롤러 메서드 작성

- [x] 인가 코드 수신 및 Access Token 발급, 사용자 정보 조회, JWT 발급 처리

- [x] Refresh Token 재발급 정책 구현

- [x] Redis에 저장된 Refresh Token 검증 및 재발급 API 구현

- [x] Access Token 만료 시 Refresh Token을 이용해 새 토큰 발급

---

8월 11일 수정 
카카오 계정 로그인 시 DB에서 회원 조회 >>  회원 조회 후 없을 때 예외처리 
예외처리로 회원가입을 유도 >> 카카오 계정 전용 회원가입 API 추가 
회원가입시 별도의 로그인없이 바로 로그인 되도록 AccessToken토큰, RefreshToken반환 

---

8월 12일 수정 
- provider 오류 수정
- RestTemplate 삭제 하려했으나 Spring Boot가 자동으로 빈을 등록해주지 않아서 직접 @Bean 등록 유지
- 일반 회원가입 Request와 소셜 회원가입 Request를 분리 (SignupRequest → SocialSignupRequest)
- kakaoOAuthService / socialLoginService 구조 중복 제거 
- 메서드 분리("회원정보 요청"과 "로그인 처리"를 별도 메서드로 나눔)
- Controller 개선 /{provider} 형태로 모든 소셜 로그인 가능하게 수정 (Kakao, Google 등 확장 가능)

기존 문제
예외가 발생하면 providerId를 프론트로 넘기지 못해서 회원가입 진행 불가
일반 회원가입 DTO(SignupRequest)를 소셜에도 사용 → 불필요한 필드 포함
소셜 로그인 정보 저장 위치 불명확

개선 내용
DB에 회원 없는 경우 처리
예외 발생 전, 소셜 로그인으로 가져온 email, provider, providerId를 SocialAccount에 임시 저장
예외 발생
프론트에서 소셜 회원가입 API를 호출하도록 유도
회원가입 API에서 provider + providerId로 SocialAccount 조회
새로운 User 생성 + SocialAccount 연결
JWT AccessToken, RefreshToken 발급
RefreshToken Redis에 저장

---

8월 13일 수정 

-  SignupCommonService.java 작성 
- 일반회원가입, 소셜회원가입  중복 로직 개별 메서드로 분리
- user엔티티 provider,providerId 삭제 
- OAuth 폴더 이동 

---

8월 14일 

- SignupCommonService.java 삭제
- 중복되는 코드 따로 메서드로 작성
- 소셜로그인 메서드 수정
- provider > providerType 이름 수정

---

## 🔍  카카오 소셜 로그인 실행하는 방법 5분기록보드 작성 

https://www.notion.so/teamsparta/2492dc3ef51480c98a46d054a02fe6ee


## 💬 기타 사항
- 관련 이슈: #265 
- 참고 자료: [링크]

---

Commit:
Feature: OAuthProvider.java 생성

- 이넘 생성

Resolves: #240

Commit:
Feature: SocialLoginResponse.java 작성

Resolves: #240

Commit:
Feature: OAuthLoginService.java작성

Resolves: #248

Commit:
Feature: OAuthLoginService.java작성

Resolves: #248

Commit:
Feature: RestTemplateConfig.java 작성

Resolves: #248

Commit:
Feature: KakaoUserInfo.java 작성

Resolves: #248

Commit:
Feature: SocialLoginResponse.java 작성

Resolves: #248

Commit:
Feature: KakaoOAuthLoginService.java 작성

- 에러코드 추가

Resolves: #248

Commit:
Feature: SocialLoginServiceFactory.java 작성

- 각 서비스 구현체에 getProvider 메서드 추가

Resolves: #248

Commit:
Feature: SocialLoginRequest.java 작성

Resolves: #248

Commit:
Feature: SocialLoginController.java 작성

Resolves: #248

Commit:
Feature: KakaoOAuthLoginService.java 작성

-토큰관련 코드 추가

Resolves: #248

Commit:
Feature: redis에 리프레시토큰 저장하는 코드

Resolves: #248

Commit:
Feature: SocialLoginController.java

- redirectToKakao 매서드 작성
  (카카오로그인페이지로 이동)

Resolves: #265

Commit:
Feature: SocialLoginController.java

- 인가 코드로 엑세스 토큰 요청

Resolves: #265

Commit:
Feature: KakaoOAuthLoginService.java

- 인가 코드로 엑세스 토큰 요청 메서드 추가

Resolves: #265

Commit:
Feature: SocialLoginServiceFactory.java

- 오류 수정

Resolves: #265

Commit:
Feature: RefreshToken 발급받기위한 dto 작성

Resolves: #265

Commit:
Feature: AuthService.java

- 리프레시 토큰 메서드 작성

Resolves: #265

Commit:
Feature: AuthController.java

- 리프레시 토큰 메서드 작성

Resolves: #265

gyogyo0212 committed:
Merge branch 'develop' into feature/#248-social-login-kakao

Commit:
Feature: application-local.yml

- 수정

Resolves: #265

Commit:
Feature: application-local.yml

- 수정

Resolves: #265

Commit:
Feature: User.java 수정

- provider ,providerId 추가

Resolves: #265

Commit:
Feature: User.java 수정

- provider ,providerId 추가

Resolves: #265

Commit:
Feature: KakaoOAuthLoginService.java

- 회원가입을 유도하기 위한 예외처리 추가

Resolves: #265

Commit:
Feature: SignupRequest.java 수정

- provider, providerId 추가

Resolves: #265

Commit:
Feature: SignupRequest.java 수정

- 소셜 회원가입 시 비밀번호 없이 넘어올 수 있으므로 비밀번호는 필수 입력에서 제외

Resolves: #265

Commit:
Feature: UserSummaryDto.java

- email 추가

Resolves: #265

Commit:
Feature: SocialLoginController.java

- 소셜 회원가입 추가 하였습니다.
 회원가입 후에 바로 토큰을 반환 해줍니다.

Resolves: #265

Commit:
Feature: SocialLoginService.java

- 소셜 회원가입 부분 따로 분리해서 작성
- 회원가입 후에 바로 토큰을 반환 해줍니다.

Resolves: #265

Commit:
feature: UserSummaryDto.java

- 빌더에 email 추가
- 이외의 파일 로그추가

resolves: #265

Commit:
Merge branch 'develop' into feature/#248-social-login-kakao

# Conflicts:
#	src/main/java/com/explorer/gabom/domain/user/entity/User.java

GyeongSe99 commented on src/main/java/com/explorer/gabom/domain/auth/controller/SocialLoginController.java:
```diff
@@ -0,0 +1,80 @@
+    // 카카오 로그인 페이지로 이동 (리다이렉트)
+    @GetMapping("/kakao")
+    public void redirectToKakao(HttpServletResponse response) throws IOException {
+        System.out.println("redirectUri = " + redirectUri);  // 이 값이 제대로 나오나요?
```
디버깅용 콘솔 출력 메소드는 삭제해주세요

kinhyunsu replied:
동의합니다.

GyeongSe99 commented on src/main/java/com/explorer/gabom/domain/auth/controller/SocialLoginController.java:
```diff
@@ -0,0 +1,80 @@
+    @GetMapping("/kakao/callback")
+    public ResponseEntity<?> kakaoCallback(@RequestParam String code) {
+        // 받은 인가 코드로 액세스 토큰 요청 및 로그인 처리 로직 작성
+        System.out.println("인가 코드: " + code);
```
디버깅용 콘솔 출력 메소드는 삭제해주세요

GyeongSe99 commented on src/main/java/com/explorer/gabom/domain/auth/controller/SocialLoginController.java:
```diff
@@ -0,0 +1,80 @@
+                "&redirect_uri=" + redirectUri +
+                "&response_type=code";
+
+        response.sendRedirect(kakaoAuthUrl);
```
response.sendRedirect(kakaoAuthUrl)말고 ResponseEntity.status(302).header("Location", url) 등의 방법으로 리다이렉트 하는건 어떤가요?

Commit:
feature: SocialLoginController.java

- 리다이렉트 주소 생성로직 서비스로 이동

resolves: #265

Commit:
feature: User.jaca

- 빌더 수정 하여 오류 수정

resolves: #265

Commit:
feature: SsoAuthToken.java 생성

-

resolves: #265

Commit:
feature: SocialLoginService.java

- 소셜로그인 공통 서비스로 변경,카카오OAuth서비스 수정 , 공통적으로 사용할 수 있게 컨트롤러 수정

resolves: #265

Commit:
feature: SocialAccount.java 작성

-

resolves: #265

Commit:
feature:

- 액세스 토큰으로 사용자 정보를 요청해서 이메일과 providerId를 추출함.
- 받은 카카오 사용자 정보로 기존 회원 조회,없으면 회원가입 유도(예외 발생),있으면 SocialAccount 테이블에 소셜 계정 정보를 저장하거나 조회,JWT 토큰을 생성해서 로그인 응답으로 반환함.

resolves: #265

Commit:
feature:

- 사용하지않는 import문 정리 코드 재정렬

resolves: #265

kinhyunsu commented on src/main/java/com/explorer/gabom/global/oauth/service/KakaoOAuthService.java:
```diff
@@ -0,0 +1,120 @@
+                String.class
+        );
+        // 여기 로그 추가 (사용자 정보 응답 출력)
+        log.debug("카카오 사용자 정보 응답: {}", response.getBody());
```
흠

Commit:
feature: SocialSignupRequest

- 기존 사용하던 요청에서 새로 생성해서 교체하였습니다.

resolves: #265

Commit:
Merge branch 'develop' into feature/#248-social-login-kakao

Commit:
feature: User.java

- provider,providerId 삭제

resolves: #265

Commit:
feature: 일반 회원가입, 소셜 회원가입 부분 수정

- 메서드 안에 중복되는 부분을 SignupCommonService.java 작성하여 코드 수정

resolves: #265

Commit:
Merge branch 'develop' into feature/#248-social-login-kakao

# Conflicts:
#	src/main/java/com/explorer/gabom/domain/user/entity/User.java

Commit:
feature: User.java

- 빌더의 status ACTIVE 삭제

resolves: #265

Commit:
feature: OAuth 폴더 위치 변경

- global.oauth를 domain.auth.oauth

resolves: #265

Commit:
feature: SignupRequest.java

- provider,providerId 삭제

resolves: #265

Commit:
feature: SignupRequest.java

- provider,providerId 삭제

resolves: #265

Commit:
Merge branch 'develop' into feature/#248-social-login-kakao

GyeongSe99 committed:
refactor: 구조 변경

Resolves: #279

Commit:
feature: jay45 님이 변경한 내용 , user.java

- user 엔티티 ofSocial, ofRegular 추가
- SignupCommonService.java 유저 생성하는 메서드 삭제
- 일반회원가입,소셜회원가입 유저생성하는 부분 수정

Resolves: #265

Commit:
feature: SignupCommonService.java 삭제

- SignupCommonService.java 삭제
- 중복되는 코드 따로 메서드로 작성
- 소셜로그인 메서드 수정

resolves: #265

Commit:
feature: provider >>providerType 이름변경

resolves: #265

Commit:
Merge branch 'develop' into feature/#248-social-login-kakao

# Conflicts:
#	src/main/java/com/explorer/gabom/domain/user/entity/User.java
#	src/main/java/com/explorer/gabom/global/exception/ErrorCode.java
#	src/main/resources/application-local.yml

Commit:
feature: socialLoginService.java 에 handleLogin작성

- socialLoginService인터페이스 이름 변경 > SocialService.java
- SocialLoginController.java 콜백 핸들러 작성

- SocialLoginService.java 회원가입 메서드명 변경

resolves: #265

Commit:
Merge branch 'develop' into feature/#248-social-login-kakao

Commit:
Merge branch 'develop' into feature/#248-social-login-kakao

Commit:
feature: 전체적인 코드 수정

resolves: #265

Commit:
feature: 전체적인 코드 수정

resolves: #265

Commit:
feature: 주석 추가

resolves: #265
